Tool for generating changelog and creating releases

pomek opened the pull request:
Fixes: #2.

pomek commented:
I introduced a lot of tests for the new code but `tasks.createRelease` and `tasks.generateChangelog` still don't have tests.

The tasks are really big and tests will be bigger. We should consider whether it can be split into smaller modules.

Reinmar commented:
> I introduced a lot of tests for the new code but tasks.createRelease and tasks.generateChangelog still don't have tests.
> 
> The tasks are really big and tests will be bigger. We should consider whether it can be split into smaller modules.

Definitely. Those functions are *far* too big. It'd be super hard to tests them now. But we'll be working on this later.

Reinmar commented:
My approach to this review will be to read the code, but I won't try to validate these tools. We'll work on that when releasing the next version of CKEditor 5.

Reinmar commented:
OK, I changed my mind :D I'll do the review now, because there's a bit too many of the small things which will be hard to polish later.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+/**
```
Let's not create such a big utils files. Split it into smaller ones, preferably a function per file (but sometimes it makes a bit more sense to group them). And do this after https://github.com/ckeditor/ckeditor5-dev/pull/8/files#r97547555, because then you'll see what's the best way to shape the code.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+			}
+		} );
+
+		delete options._;
```
Please remove this or comment it and write a test for it.

Hint: Since it'd be a waste of time to test it it means that this code makes no sense and could be removed :D

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+	},
+
+	/**
+	 * @returns {Promise}
```
If the only thing this function does is reading a file from the system, please get rid of the promise. I've been writing this on and on in many PRs – use synchronous versions of Node.js methods.

PS. It took me 15s to try to understand what this function does. And I'm still unsure, which is exactly why KISS!

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+	 * @param {String} content
+	 * @returns {Promise}
+	 */
+	saveChangelog( content ) {
```
The same as above.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+	 * @returns {Promise}
+	 */
+	saveChangelog( content ) {
+		const changelogFile = path.resolve( utils.changelogFile );
```
The same as above.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+	 * @returns {Promise}
+	 */
+	getCurrentChangelog() {
+		const changelogFile = path.resolve( utils.changelogFile );
```
This doesn't seem to be necessary.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/changelog.js:
```diff
@@ -0,0 +1,249 @@
+	 * @param {String} options.debug Whether to display additional logs.
+	 * @returns {Promise}
+	 */
+	createGithubRelease( token, options ) {
```
This isn't a util. This is a specific task. It doesn't mean that it should end in the tasks.js file, but that it shouldn't be in the utils :P.

The utils hole is the last place where you want a file to end. And it must meet certain requirements, like having no requirements of its own other than other utils.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/exec-on-dependencies.js:
```diff
@@ -0,0 +1,53 @@
+/**
```
Wrong file and function names. Should be `executeOnDependencies()`

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/exec-on-dependencies.js:
```diff
@@ -0,0 +1,53 @@
+	}
+
+	return promise.then( () => {
+		process.chdir( currentCwd );
```
Why is it resetting the cwd if it's not changing it?

pomek replied:
`functionToExecute` can change the cwd.

Reinmar replied:
So the function itself should fix the dir.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/exec-on-dependencies.js:
```diff
@@ -0,0 +1,53 @@
+const { workspace } = require( '@ckeditor/ckeditor5-dev-utils' );
+
+/**
+ * The module allows running a function on their locally installed dependencies.
```
It's not a module.

Reinmar commented on packages/ckeditor5-dev-env/lib/utils/proposed-deps-version.js:
```diff
@@ -0,0 +1,43 @@
+/**
```
Wrong file and function names. `proposedDependenciesVersions()`

Reinmar commented on packages/ckeditor5-dev-env/tests/fixtures/package.json:
```diff
@@ -0,0 +1,8 @@
+{
```
Don't assume that there will be just on fixture project needed. Wrap it in `basic/` or similar directory.

Reinmar commented on packages/ckeditor5-dev-env/tests/fixtures/package.json:
```diff
@@ -0,0 +1,8 @@
+{
+	"version": "3.1.2",
+	"dependencies": {
+		"ckeditor5-core": "ckeditor/ckeditor5-core",
```
Wrong names and versions (we're using scoped and versioned deps already).

pomek replied:
This code has been written a few weeks before we started to use the scoped :D

Reinmar commented on packages/ckeditor5-dev-env/tests/fixtures/packages/ckeditor5-core/package.json:
```diff
@@ -0,0 +1,3 @@
+{
+	"name": "ckeditor5-core"
```
Should be scoped.

pomek replied:
This is a fixture which checks whether the correct name is passed to function.

pomek commented:
Because `tasks.createRelease` is split into small modules, I wrote tests for it.

Unfortunately, I had problems with tests for `tasks.generateChangelog`. I couldn't create a stream mock which will be compatible with our `stream.noop()` helper.

Reinmar commented:
OK, I made a review. I haven't been able to check the whole code, because it's too much of it, but I checked how the tool works and I fixed some issues. I also aligned the setup to our needs.

However, I found one critical issue. According to the spec, the tool should only read commits made directly to the master branch, without going into commits done in in ticket branches (in those cases it should only check the merge commits). How the tool works now is completely the opposite – it checks every commit, including ticket branches, but it ignores merge commits. This needs to be fixed before we close this PR.

Reinmar commented:
![image](https://cloud.githubusercontent.com/assets/156149/22380468/e4ce0364-e4bd-11e6-8189-f93c5bae4ba3.png)

```
(master 1dbd0e5) p@m /workspace/test-releases/ckeditor5> git log --first-parent master
commit 1dbd0e5da426974902cb5f973b4fd6e3e0918cce
Author: Piotrek Koszuliński <pkoszulinski@gmail.com>
Date:   Fri Jan 27 17:32:14 2017 +0100

    Removed unreleased changelog entries.

commit 6038d4c0cdfe1571c9bb4803471f79b536eb9b28
Merge: bd2e76b 56f0ebe
Author: Piotrek Koszuliński <pkoszulinski@gmail.com>
Date:   Fri Jan 27 17:30:18 2017 +0100

    Fix: This should count.

commit bd2e76b678de3ee86a2a10ce52d3431ee067d37b
Author: Piotrek Koszuliński <pkoszulinski@gmail.com>
Date:   Fri Jan 27 17:12:41 2017 +0100

    Fix: Aligned gulpfile to changes in ckeditor5-dev-env.

commit c9e65fc60a84504ad8ca51d758a66aaa3bee3edd
Author: Piotrek Koszuliński <pkoszulinski@gmail.com>
Date:   Fri Jan 27 16:14:37 2017 +0100

    Docs: Changelog.

```

Reinmar commented:
I'm only a bit unsure how to process merge commits. In the example above I deleted the original message, which normally look like this:

```
commit c41f115d11a6a092f82d42c1bf502b7f613d5d77
Merge: 1ea8f69 03dcb6b
Author: Piotrek Koszuliński <pkoszulinski@gmail.com>
Date:   Fri Dec 16 14:39:11 2016 +0100

    Merge pull request #382 from ckeditor/t/381

    Remove unnecessary tasks.
```

The "Merge pull request ..." line is pretty useful and I'd rather like to keep it, which means that our  merge commits should look like this:

```
Merge pull request #382 from ckeditor/t/381

Fix: This should be the message for the changelog. Fixes #123.
```

Let's see how to pass it to conventional changelog ;>

pomek commented:
I fixed it.

For the following tree:

![image](https://cloud.githubusercontent.com/assets/2270764/22417633/83376a7c-e6d5-11e6-9777-58145230244a.png)

I tried to modify the configuration for `gitRawCommits`.

### The first one - `{ merges: false }` (by default):

![image](https://cloud.githubusercontent.com/assets/2270764/22417681/bb66d64e-e6d5-11e6-8abb-81f868389ab9.png)

**Merge commits are skipped but ticket commits are available.**

### The second one - `{ merges: true }`:

![image](https://cloud.githubusercontent.com/assets/2270764/22417716/dfc08d96-e6d5-11e6-9b0f-98d315bc2455.png)

**Merge commits are available and ticket commits are hidden.**

---

I've created another branch, pushed two commits to them:

![image](https://cloud.githubusercontent.com/assets/2270764/22417795/54244b5a-e6d6-11e6-8030-aa90ca488d8b.png)

Then I merged this branch into the master. I updated the message of merge commit:

![image](https://cloud.githubusercontent.com/assets/2270764/22417807/661d2de0-e6d6-11e6-8b25-85848008f76a.png)

And then I generated the changelog:

![image](https://cloud.githubusercontent.com/assets/2270764/22417822/79b659a8-e6d6-11e6-83e2-84f32deec14a.png)

![image](https://cloud.githubusercontent.com/assets/2270764/22417833/83ecd88e-e6d6-11e6-8bd0-e4b7a90aa05d.png)

---

So, the ticket commits are skipped, merge commits are parsed and it works as we want.

pomek commented:
Finally, I did it. Git takes commits from a master and also merge commits into the master.

![image](https://cloud.githubusercontent.com/assets/2270764/22423978/acf74728-e6f4-11e6-909b-1abe4ebdaeb6.png)

Reinmar commented on packages/ckeditor5-dev-env/lib/release-tools/changelog/writer-options.js:
```diff
@@ -47,6 +48,16 @@ module.exports = {
 // - filters out the commit if it should not be visible in the changelog,
 // - makes links to issues and user's profiles on GitHub.
 function transformCommit( commit ) {
+	if ( commit.header.startsWith( 'Merge pull request' ) ) {
```
What about merges done from the console? The message is just "Merge ...", so I'll change this check to `startsWith( 'Merge' )`

pomek replied:
Done.

Reinmar commented:
LGTM now. I have some followups to report, but the tasks do their job pretty much correctly now. Good job!
